File: Medium/2353. Design a Food Rating System.java
class Food implements Comparable<Food> {
    public String name;
    public String cuisine;
    public int rating;

    public Food(String name, String cuisine, int rating) {
        this.name = name;
        this.cuisine = cuisine;
        this.rating = rating;
    }

    @Override
    public int compareTo(Food other) {
        if (this.rating == other.rating) {
            return this.name.compareTo(other.name);
        }
        return -1 * Integer.compare(this.rating, other.rating);
    }
}

class FoodRatings {
    private Map<String, Food> nameToFood;
    private Map<String, PriorityQueue<Food>> cuisineToFoods;

    public FoodRatings(String[] foods, String[] cuisines, int[] ratings) {
        nameToFood = new HashMap<>();
        cuisineToFoods = new HashMap<>();
        for (int i = 0; i < foods.length; i++) {
            Food food = new Food(foods[i], cuisines[i], ratings[i]);
            nameToFood.put(foods[i], food);
            cuisineToFoods.computeIfAbsent(cuisines[i], k -> new PriorityQueue<>()).offer(food);
        }
    }

    public void changeRating(String food, int newRating) {
        Food oldFood = nameToFood.get(food);
        oldFood.name = "";
        Food newFood = new Food(food, nameToFood.get(food).cuisine, newRating);
        nameToFood.put(food, newFood);
        cuisineToFoods.get(newFood.cuisine).offer(newFood);
    }

    public String highestRated(String cuisine) {
        String highestOne = cuisineToFoods.get(cuisine).peek().name;
        while (highestOne.equals("")) {
            cuisineToFoods.get(cuisine).poll();
            highestOne = cuisineToFoods.get(cuisine).peek().name;
        }
        return highestOne;
    }
}

/**
 * Your FoodRatings object will be instantiated and called as such:
 * FoodRatings obj = new FoodRatings(foods, cuisines, ratings);
 * obj.changeRating(food,newRating);
 * String param_2 = obj.highestRated(cuisine);
 */

File: Medium/2353. Design a Food Rating System.py
from collections import defaultdict
import heapq
from typing import List


class FoodRatings:
    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        self.food_to_cuisine = {}
        self.food_to_rating = {}
        self.cuisine_to_highest = defaultdict(list)
        for food, cuisine, rating in zip(foods, cuisines, ratings):
            self.food_to_cuisine[food] = cuisine
            self.food_to_rating[food] = rating
            heapq.heappush(self.cuisine_to_highest[cuisine], (-rating, food))

    def changeRating(self, food: str, newRating: int) -> None:
        self.food_to_rating[food] = newRating
        heapq.heappush(self.cuisine_to_highest[self.food_to_cuisine[food]], (-newRating, food))

    def highestRated(self, cuisine: str) -> str:
        heap = self.cuisine_to_highest[cuisine] 
        while heap:
            highest, food = heap[0]
            if -highest == self.food_to_rating[food]:
                return food
            heapq.heappop(heap)
        return ""


# Your FoodRatings object will be instantiated and called as such:
# obj = FoodRatings(foods, cuisines, ratings)
# obj.changeRating(food,newRating)
# param_2 = obj.highestRated(cuisine)